src/core/components/pin/pin-manager.js

```diff
@@ -0,0 +1,304 @@
+
+          cids
+            .map(cid => cid.toString())
+            // recursive pins pre-empt indirect pins
+            .filter(key => !this.recursivePins.has(key))
+            .forEach(key => indirectKeys.add(key))
+
+          cb()
+        })
+      })
+    }, (err) => {
+      if (err) { return callback(err) }
+      callback(null, Array.from(indirectKeys))
+    })
+  }
+
+  // Encode and write pin key sets to the datastore:
+  // a DAGLink for each of the recursive and direct pinsets
+  // a DAGNode holding those as DAGLinks, a kind of root pin
+  flushPins (callback) {
+    let dLink, rLink, root
+    series([
+      // create a DAGLink to the node with direct pins
+      cb => waterfall([
+        cb => this.pinset.storeSet(this.directKeys(), cb),
+        ({ node, cid }, cb) => {
+          try {
+            cb(null, new DAGLink(PinTypes.direct, node.size, cid))
+          } catch (err) {
+            cb(err)
+          }
+        },
+        (link, cb) => { dLink = link; cb(null) }
+      ], cb),
+
+      // create a DAGLink to the node with recursive pins
+      cb => waterfall([
+        cb => this.pinset.storeSet(this.recursiveKeys(), cb),
+        ({ node, cid }, cb) => {
+          try {
+            cb(null, new DAGLink(PinTypes.recursive, node.size, cid))
+          } catch (err) {
+            cb(err)
+          }
+        },
+        (link, cb) => { rLink = link; cb(null) }
+      ], cb),
+
+      // the pin-set nodes link to a special 'empty' node, so make sure it exists
+      cb => {
+        let empty
+
+        try {
+          empty = DAGNode.create(Buffer.alloc(0))
+        } catch (err) {
+          return cb(err)
+        }
+
+        this.dag.put(empty, {
+          version: 0,
+          format: multicodec.DAG_PB,
+          hashAlg: multicodec.SHA2_256,
+          preload: false
+        }, cb)
+      },
+
+      // create a root node with DAGLinks to the direct and recursive DAGs
+      cb => {
+        let node
+
+        try {
+          node = DAGNode.create(Buffer.alloc(0), [dLink, rLink])
+        } catch (err) {
+          return cb(err)
+        }
+
+        root = node
+        this.dag.put(root, {
+          version: 0,
+          format: multicodec.DAG_PB,
+          hashAlg: multicodec.SHA2_256,
+          preload: false
+        }, (err, cid) => {
+          if (!err) {
+            root.multihash = cid.buffer
+          }
+          cb(err)
+        })
+      },
+
+      // hack for CLI tests
+      cb => this.repo.closed ? this.repo.open(cb) : cb(null, null),
+
+      // save root to datastore under a consistent key
+      cb => this.repo.datastore.put(PIN_DS_KEY, root.multihash, cb)
+    ], (err, res) => {
+      if (err) { return callback(err) }
+      this.log(`Flushed pins with root: ${root}`)
+      return callback(null, root)
+    })
+  }
+
+  load (callback) {
+    waterfall([
+      // hack for CLI tests
+      (cb) => this.repo.closed ? this.repo.datastore.open(cb) : cb(null, null),
```

Comment:
do we need this hack?

Reply:
And if it is, explain what the hack is and why it was necessary

Reply:
Good question - this was in the original code that I [refactored over from pin.js](https://github.com/ipfs/js-ipfs/blob/master/src/core/components/pin.js#L159), do you know why it might be there @daviddias?

Reply:
It seems like if I remove it the tests work fine, so I think I will just go ahead and do that. Good catch @hugomrdias 👍